api/app/tests/fba/test_fba_error_random_sample.py

```diff
@@ -0,0 +1,239 @@
+"""
+Unit tests for fire behavour calculator.
+"""
+from datetime import datetime, timezone as dt_tz
+import random
+from typing import Final
+import logging
+from pytest_bdd import scenario, given, then
+from app import configure_logging
+from app.utils.time import get_hour_20_from_date
+from app.utils.fba_calculator import calculate_fire_behaviour_advisory, FBACalculatorWeatherStation
+from app.utils.redapp import FBPCalculateStatisticsCOM
+from app.utils.cffdrs import initial_spread_index, bui_calc
+import pytest
+
+
+configure_logging()
+
+logger = logging.getLogger(__name__)
+
+
+def _str2float(value: str):
+    if value == 'None':
+        return None
+    return float(value)
+
+
+def _random_date():
+    start = datetime.fromisoformat('2021-01-01')
+    end = datetime.fromisoformat('2021-12-31')
+    return datetime.fromtimestamp(random.uniform(start.timestamp(), end.timestamp()), tz=dt_tz.utc)
+
+
+def relative_error(actual: float, expected: float, precision: int = 2):
+    """ calculate the relative error between two values - default to precision of 2"""
+    actual = round(actual, precision)
+    expected = round(expected, precision)
+    if actual == expected:
+        return 0
+    if expected == 0:
+        # Can't divide by 0! Taking this as a 100% difference
+        return 1
+    return abs((actual-expected)/expected)
+
+
+acceptable_margin_of_error: Final = 0.01
+
+
+def check_metric(metric: str,
+                 fuel_type: str,
+                 python_value: float,
+                 comparison_value: float,
+                 metric_error_margin: float, note: str = None):
+    """ Check relative error of a metric """
+    # logging with %s became unreadable:
+    # pylint: disable=logging-fstring-interpolation
+    if comparison_value < 0:
+        logger.warning('Skipping %s! (%s) - note: %s', metric, comparison_value, note)
+    else:
+        assert python_value >= 0
+        error = relative_error(python_value, comparison_value)
+        absolute_error = abs(python_value - comparison_value)
+        # logger.info(
+        #     f'{fuel_type}: Python {python_value}, {metric} {comparison_value} ; error: {error}')
+        if error > acceptable_margin_of_error:
+            logger.error('%s %s relative error %s > %s! (python: %s, expected: %s)',
+                         fuel_type, metric, error, acceptable_margin_of_error, python_value, comparison_value)
+        if metric_error_margin > 0.01:
+            logger.debug('%s: The acceptable margin of error (%s) for %s is set too high',
+                         fuel_type, metric_error_margin, metric)
+        if error > metric_error_margin:
+            # ok - fine, but for small numbers this isn't a big deal, so let's check the absolure difference.
+            if absolute_error < 0.01:
```

Comment:
what's your rationale for 0.01 in `absolute_error < 0.01`? I would've more likely expected 0.1, since in the frontend we round to only 1 decimal place

Reply:
Arbitrary decision really; I just want to not fail when even though the relative error is big, we're actually talking tiny numbers. 
That said, with CFB, 0.01 would be a full percentage point difference. We multiply CFB by 100, so 0.01 works for CFB, but maybe not as well for others.

Reply:
ahhh right. I forgot that we're working with such a small range in CFB